### PR TITLE
i18n: use tsc to collect UIStrings

### PR DESCRIPTION
**Pull request by brendankenny**
moves from using esprima to the typescript compiler to extract UIStrings from our files. Discussed this with @exterkamp ahead of time, though this PR grew a little in the telling. Main benefits:
- eliminates an aging and [seldom updated](https://www.npmjs.com/package/esprima?activeTab=versions) dependency for one that we keep updated and we already know can parse our source
- tsc automatically extracts jsdoc strings including putting `@` tags in separate vars, eliminating one parse step and having to keep track of comment start/end positions
- automatically does some format corrections, e.g. removes `*` and starting whitespace from jsdoc strings, joins across newlines, etc
- is actually typed :)

also reorganized a bit so the parsing/ast walking is split off in a separate function, allowing testing to be a bit easier (and laying the groundwork for exposing as a utility for third-party plugins, etc, to use if they wish)

### CODE REVIEW

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 1]
```diff
@@ -12,14 +12,15 @@ const fs = require('fs');
 const glob = require('glob');
 const path = require('path');
 const assert = require('assert');
-const esprima = require('esprima');
+const tsc = require('typescript');
 const collectAndBakeCtcStrings = require('./bake-ctc-to-lhl.js');
 
 const LH_ROOT = path.join(__dirname, '../../../');
-const UISTRINGS_REGEX = /UIStrings = (.|\s)*?\};\n/im;
+const UISTRINGS_REGEX = /UIStrings = .*?\};\n/s;
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: as this has evolved we no longer actually use the capture group, the case insensitivity, or the multiline flag. The `s` flag allows `.` to encapsulate `/s` as well.

**Comment by brendankenny** [comment]
the changes to `collectAllStringsInDir` are easier to read if you hide whitespace, as a lot of it is removing a level of indentation

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 2]
```diff
@@ -375,6 +389,60 @@ function createPsuedoLocaleStrings(messages) {
...
+  const uiStringsStatement = tsAst.statements[0];
+  if (tsAst.statements.length !== 1) throw extractionError;
+  if (!tsc.isVariableStatement(uiStringsStatement)) throw extractionError;
+
+  const uiStringsDeclaration = uiStringsStatement.declarationList.declarations[0];
+  if (!tsc.isVariableDeclaration(uiStringsDeclaration)) throw extractionError;
+  if (getIdentifier(uiStringsDeclaration) !== 'UIStrings') throw extractionError;
+
+  const uiStringsObject = uiStringsDeclaration.initializer;
+  if (!uiStringsObject || !tsc.isObjectLiteralExpression(uiStringsObject)) throw extractionError;
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: we could alternatively drop a lot of the 9 lines above, set
```js
let uiStringsObject;
try {
  // @ts-ignore
  uiStringsObject = tsAst.statements[0].declarationList.declarations[0].initializer;
  if (tsc.isObjectLiteralExpression(uiStringsObject)) throw new Error('bad');
} catch (e) {
  throw new Error('UIStrings was not extracted correctly by the collect-strings regex.');
}
```
The type narrowing as we walk down is of questionable value since
a) we created the parsed string ourselves
b) we throw on any and all assumption failures anyway
c) meanwhile it looks more important than it is

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 3]
```diff
@@ -375,6 +389,60 @@ function createPsuedoLocaleStrings(messages) {
...
+
+  /** @type {Record<string, ParsedUIString>} */
+  const parsedMessages = {};
+
+  for (const property of uiStringsObject.properties) {
+    const key = getIdentifier(property);
+
+    // Use live message to avoid having to e.g. concat strings broken into parts.
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: I suspect this is why @patrickhulce did this in the first place, but I'm not sure :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 4]
```diff
@@ -402,66 +470,54 @@ function collectAllStringsInDir(dir) {
...
+      throw new Error('UIStrings exported but no definition found');
     }
 
-    if (exportsUIStrings && !regexMatches) {
-      throw new Error('UIStrings exported but no definition found');
+    if (!exportedUIStrings) {
+      throw new Error('UIStrings defined but not exported');
     }
 
-    // @ts-ignore regex just matched
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: rearranged above to avoid this

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 5]
```diff
@@ -513,7 +569,7 @@ if (require.main === module) {
 }
 
 module.exports = {
-  computeDescription,
+  parseUIStrings,
   createPsuedoLocaleStrings,
-  convertMessageToPlaceholders: convertMessageToCtc,
+  convertMessageToCtc,
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: it didn't make sense to rename this just for exporting for tests, but missed in the review for #9114 :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 6]
```diff
@@ -402,66 +470,54 @@ function collectAllStringsInDir(dir) {
...
+      throw new Error('UIStrings defined but not exported');
     }
 
-    // @ts-ignore regex just matched
-    const justUIStrings = 'const ' + content.match(UISTRINGS_REGEX)[0];
     // just parse the UIStrings substring to avoid ES version issues, save time, etc
-    // @ts-ignore - esprima's type definition is supremely lacking
-    const ast = esprima.parse(justUIStrings, {comment: true, range: true});
-
-    for (const stmt of ast.body) {
```
Comment (brendankenny) [inline lighthouse-core/scripts/i18n/collect-strings.js]: there's only ever one `stmt` (if we did our extraction correctly), so no need to loop :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 7]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+  }
 
+  if (ast.comment) {
     // The entire comment is the description, so return everything.
-    return {description};
+    return {description: singleLineJsDocComment(ast.comment), examples: {}};
   }
-  throw Error(`No Description for message "${value}"`);
+
+  throw Error(`No Description for message "${message}"`);
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]:
```suggestion
  throw Error(`Missing description comment for message "${message}"`);
```

while we're in here :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 8]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+}
+
+/**
+ * Line breaks within a jsdoc comment should always be replaceable with a space.
+ * @param {string=} comment
+ * @return {string}
+ */
+function singleLineJsDocComment(comment = '') {
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]: could I offer you any of...

- `coerceToSingleLine`
- `coerceToSingleLineAndTrim`
- `flattenToSingleLineJsDocComment`
- `replaceNewLines`

? :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 9]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+  return comment.replace(/\n/g, ' ').trim();
+}
+
+/**
+ * Parses a tag of the form `@example {exampleValue} placeholderName`.
+ * @param {string} rawExample
+ * @return {{placeholderName: string, exampleValue: string}}
+ */
+function parseExample(rawExample) {
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]: would it be too crazy to call this `parseAtExample`? :)

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 10]
```diff
@@ -375,6 +389,60 @@ function createPsuedoLocaleStrings(messages) {
...
+
+/**
+ * @param {string} sourceStr String of the form 'const UIStrings = {...}'.
+ * @param {Record<string, string>} liveUIStrings The actual imported UIStrings object.
+ * @return {Record<string, ParsedUIString>}
+ */
+function parseUIStrings(sourceStr, liveUIStrings) {
+  const tsAst = tsc.createSourceFile('uistrings', sourceStr, tsc.ScriptTarget.ES2019, true, tsc.ScriptKind.JS);
+
+  const extractionError = new Error('UIStrings was not extracted correctly by the collect-strings regex.');
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]:
```suggestion
  const extractionError = new Error('UIStrings declaration was not extracted correctly by the collect-strings regex.');
```

is this accurate?

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/scripts/i18n/collect-strings.js]: eh I kinda like the way it explains the structure of everything, so I'm for keeping as-is now that we have it 👍

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/scripts/i18n/collect-strings.js]: yes, precisely :) I think I got like 20% of the way done until I hit the first file that split a line and was like f' it let's just `require` 😆

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 11]
```diff
@@ -375,6 +389,60 @@ function createPsuedoLocaleStrings(messages) {
...
+
+  for (const property of uiStringsObject.properties) {
+    const key = getIdentifier(property);
+
+    // Use live message to avoid having to e.g. concat strings broken into parts.
+    const message = liveUIStrings[key];
+
+    // @ts-ignore - Not part of the public tsc interface yet.
+    const jsDocComments = tsc.getJSDocCommentsAndTags(property);
+    if (jsDocComments.length === 0) throw new Error(`No Description for message "${message}"`);
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]: can we get rid of the message inside `computeDescription` then and make this the "Missing description comment" one?

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js [hunk 12]
```diff
@@ -136,17 +163,21 @@ describe('Compute Description', () => {
...
-    const ast = esprima.parse(justUIStrings, {comment: true, range: true});
-
-    const stmt = ast.body[0];
-    const prop = stmt.declarations[0].init.properties[0];
-    const res = collect.computeDescription(ast, prop, 'Hello World {variable} {variable2}', 0);
-    expect(res.description).toBe('Tagged description for Hello World.');
-    expect(res.examples['variable']).toBe('Variable example.');
-    expect(res.examples['variable2']).toBe('Variable2 example.');
+    expect(res).toEqual({
+      messageStr: {
```
Comment (patrickhulce) [inline lighthouse-core/test/scripts/i18n/collect-strings-test.js]: is there significance to all of these being called `messageStr`?

Comment (brendankenny) [reply lighthouse-core/test/scripts/i18n/collect-strings-test.js]: > is there significance to all of these being called `messageStr`?

no, I just kept confusing myself with `message` as a property on the compared object *and* as a property on the `message` property. If this renames it it should probably rename it something totally different instead of being super lazy like this

**Comment by patrickhulce** [comment]
> eliminates an aging and seldom updated dependency

Kinda crazy that esprima has more than double weekly downloads than typescript still o.O

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 7 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > while we're in here :)

done

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 8 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: `coerceToSingleLineAndTrim` it is!

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 9 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > would it be too crazy to call this `parseAtExample`? :)

lol done

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 10 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > is this accurate?

makes sense

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 11 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > can we get rid of the message inside `computeDescription` then and make this the "Missing description comment" one?

hmm, do you mean like check that either `ast.comment` or `ast.tags` is defined here? Or output an empty `description` from `computeDescription` and throw here if it's the empty string?

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 13]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
-    return {description};
+    return {description: coerceToSingleLineAndTrim(ast.comment), examples: {}};
   }
-  throw Error(`No Description for message "${value}"`);
+
+  throw Error(`Missing description comment for message "${message}"`);
+}
+
+/**
+ * Line breaks within a jsdoc comment should always be replaceable with a space.
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: :heart_eyes: getting multi line for free

I feel like this function description is worded more like a rule of the code, and not what the function does. I.e
```suggestion
Collapses a jsdoc comment into a single line and trims whitespace.
```

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 14]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+  if (ast.tags) {
+    // This is a complex description with description and examples.
+    let description = '';
+    /** @type {Record<string, string>} */
+    const examples = {};
+
+    for (const tag of ast.tags) {
+      const comment = coerceToSingleLineAndTrim(tag.comment);
+
+      if (tag.tagName.text === 'description') {
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: Should this throw if we encounter a non-standard tag?

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 15]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+function coerceToSingleLineAndTrim(comment = '') {
+  return comment.replace(/\n/g, ' ').trim();
+}
+
+/**
+ * Parses a tag of the form `@example {exampleValue} placeholderName`.
+ * @param {string} rawExample
+ * @return {{placeholderName: string, exampleValue: string}}
+ */
+function parseAtExample(rawExample) {
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: I get the `parse@Example` notation, but I found it a bit confusing at first, maybe `parseExampleTag` or `parseExampleJsdoc`?

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 16]
```diff
@@ -30,48 +31,61 @@ const ignoredPathComponents = [
...
+  const match = rawExample.match(/^{(?<exampleValue>[^}]+)} (?<placeholderName>.+)$/);
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: So, at this point, the raw example is in the form "{exampleValue} placeholderName"?  The `@`tag is gone? Seems like this wouldn't be the case based on the function description.

Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: We should use named groups more often...

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 17]
```diff
@@ -375,6 +389,60 @@ function createPsuedoLocaleStrings(messages) {
   return psuedoLocalizedStrings;
 }
 
+/**
+ * @param {import('typescript').NamedDeclaration} node
+ * @return {string}
+ */
+function getIdentifier(node) {
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: Explanation of what this does?  Seems like a tsc helper method to get the text for an ast property?

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js [hunk 18]
```diff
@@ -8,68 +8,65 @@
...
+  return Function(`'use strict'; ${justUIStrings} return UIStrings;`)();
+}
+
+describe('parseUIStrings', () => {
   it('collects description', () => {
     const justUIStrings =
     `const UIStrings = {
         /** Description for Hello World. */
-        message: 'Hello World',
+        exampleString: 'Hello World',
```
Comment (exterkamp) [inline lighthouse-core/test/scripts/i18n/collect-strings-test.js]: 1/2 of the ` strings in this file are indented 1 too many levels, like this one.  Can we unify all these examples on 6 spaces?
```suggestion
      exampleString: 'Hello World',
```

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js [hunk 19]
```diff
@@ -79,34 +76,60 @@ describe('Compute Description', () => {
...
+      exampleString: {
+        message: 'Hello World',
+        description: 'Tagged description for Hello World.',
+        examples: {},
+      },
+    });
   });
 
-  it.skip('collects complex multi-line description', () => {
+  it('collects complex multi-line description', () => {
```
Comment (exterkamp) [inline lighthouse-core/test/scripts/i18n/collect-strings-test.js]: :heart: these tests.  Can we get one with a multi-line description and an example too?

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js [hunk 20]
```diff
@@ -155,44 +186,137 @@ describe('Compute Description', () => {
...
+        /** 
+         * @description Some description.
+         * @example {missingPlaceholdername}
+         */
+        exampleString: 'Hello World',
+    };`;
+    const liveUIStrings = evalJustUIStrings(justUIStrings);
+
+    expect(() => collect.parseUIStrings(justUIStrings, liveUIStrings))
+      .toThrow(/Incorrectly formatted @example/);
```
Comment (exterkamp) [inline lighthouse-core/test/scripts/i18n/collect-strings-test.js]: Is this error descriptive enough inside a stack trace?
image

I think it needs more context around it, what message was bad, or what `@example` text it was iterating over.  This would help immediately know what is wrong.

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js [hunk 21]
```diff
@@ -155,44 +186,137 @@ describe('Compute Description', () => {
...
+        /** 
+         * @description Some description.
+         * @example placeholderName
+         */
+        exampleString: 'Hello World',
+    };`;
+    const liveUIStrings = evalJustUIStrings(justUIStrings);
+
+    expect(() => collect.parseUIStrings(justUIStrings, liveUIStrings))
+      .toThrow(/Incorrectly formatted @example/);
```
Comment (exterkamp) [inline lighthouse-core/test/scripts/i18n/collect-strings-test.js]: Same as above.

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 22]
```diff
@@ -402,66 +470,54 @@ function collectAllStringsInDir(dir) {
...
     const exportVars = require(absolutePath);
-    const regexMatches = UISTRINGS_REGEX.test(content);
-    const exportsUIStrings = !!exportVars.UIStrings;
-    if (!regexMatches && !exportsUIStrings) continue;
+    const regexMatch = content.match(UISTRINGS_REGEX);
+    const exportedUIStrings = exportVars.UIStrings;
 
-    if (regexMatches && !exportsUIStrings) {
-      throw new Error('UIStrings defined but not exported');
+    if (!regexMatch) {
```
Comment (exterkamp) [inline lighthouse-core/scripts/i18n/collect-strings.js]: Dunno if it's worth it, but can this be a XOR then a ternary on which error to throw? Ignore if it's pointless, but this kind of boolean logic is deceptively simple, but yet hard to follow.

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 11 above)
Comment (patrickhulce) [reply lighthouse-core/scripts/i18n/collect-strings.js]: yeah I meant this is a duplicate case and error message of the fallback in `computeDescription`, could `computeDescription` accept `ast|undefined` and do the check there?

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 13 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > I feel like this function description is worded more like a rule of the code, and not what the function does

fair, fixed

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 16 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > So, at this point, the raw example is in the form "{exampleValue} placeholderName"? The `@`tag is gone? Seems like this wouldn't be the case based on the function description.

good call. I reworded to make it clearer what it was parsing, but also left the `@example` context so it's not pretending to not be used for that specific purpose.

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 17 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > Explanation of what this does? Seems like a tsc helper method to get the text for an ast property?

done

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 22 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > Dunno if it's worth it, but can this be a XOR then a ternary on which error to throw? Ignore if it's pointless, but this kind of boolean logic is deceptively simple, but yet hard to follow.

that sounds more complicated, but I added a comment and an error message tweak to try to make it clearer :)

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js (hunk 18 above)
Comment (brendankenny) [reply lighthouse-core/test/scripts/i18n/collect-strings-test.js]: > 1/2 of the ` strings in this file are indented 1 too many levels, like this one. Can we unify all these examples on 6 spaces?

didn't you write these lines :P

fixed

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js (hunk 19 above)
Comment (brendankenny) [reply lighthouse-core/test/scripts/i18n/collect-strings-test.js]: > Can we get one with a multi-line description and an example too?

done

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js (hunk 20 above)
Comment (brendankenny) [reply lighthouse-core/test/scripts/i18n/collect-strings-test.js]: > I think it needs more context around it, what message was bad, or what @example text it was iterating over. This would help immediately know what is wrong.

good point. added example string to error message

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 11 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > could `computeDescription` accept `ast|undefined` and do the check there?

got it

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 23]
```diff
@@ -30,48 +31,69 @@ const ignoredPathComponents = [
...
+    for (const tag of ast.tags) {
+      const comment = coerceToSingleLineAndTrim(tag.comment);
+
+      if (tag.tagName.text === 'description') {
+        description = comment;
+      } else if (tag.tagName.text === 'example') {
+        const {placeholderName, exampleValue} = parseExampleJsDoc(comment);
+        examples[placeholderName] = exampleValue;
+      } else {
+        throw new Error('bad tag' + tag.tagName.text);
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]:
```suggestion
        throw new Error(`Unexpected tagName "@${tag.tagName.text}"`);
```

File: lighthouse-core/scripts/i18n/collect-strings.js [hunk 24]
```diff
@@ -30,48 +31,69 @@ const ignoredPathComponents = [
...
+}
+
+/**
+ * Collapses a jsdoc comment into a single line and trims whitespace.
+ * @param {string=} comment
+ * @return {string}
+ */
+function coerceToSingleLineAndTrim(comment = '') {
+  // Line breaks within a jsdoc comment should always be replaceable with a space.
+  return comment.replace(/\n/g, ' ').trim();
```
Comment (patrickhulce) [inline lighthouse-core/scripts/i18n/collect-strings.js]: FWIW I tend to go with `/\n+/g, ' '` if i'm turning into single line in other contexts, not sure if this is different

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 23 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: Haha, whoops, I threw that in there to see how hard it was to accidentally trigger that case with reasonable typos and then I guess accidentally committed it. Maybe we'll go with throwing in that case after all :)

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 14 above)
Comment (exterkamp) [reply lighthouse-core/scripts/i18n/collect-strings.js]: That's fair.  I'm not sure I'm convinced, but I guess there isn't much harm in letting new tags roam free.

File: lighthouse-core/test/scripts/i18n/collect-strings-test.js (hunk 18 above)
Comment (exterkamp) [reply lighthouse-core/test/scripts/i18n/collect-strings-test.js]: I don't recall ever seeing these files before in my life.

File: lighthouse-core/scripts/i18n/collect-strings.js (hunk 24 above)
Comment (brendankenny) [reply lighthouse-core/scripts/i18n/collect-strings.js]: > FWIW I tend to go with `/\n+/g, ' '` if i'm turning into single line in other contexts, not sure if this is different

it might be stretching interpreting a string with multiple line breaks in it as still the same jsdoc string but I think it's the only thing that makes sense here, so, sounds good :)
